? arm/sunxi/sun8i_v3s_gpio.c
Index: arm/sunxi/files.sunxi
===================================================================
RCS file: /cvsroot/src/sys/arch/arm/sunxi/files.sunxi,v
retrieving revision 1.70
diff -b -u -r1.70 files.sunxi
--- arm/sunxi/files.sunxi	7 Nov 2021 17:13:12 -0000	1.70
+++ arm/sunxi/files.sunxi	14 Jun 2022 10:13:01 -0000
@@ -137,6 +137,7 @@
 file	arch/arm/sunxi/sun7i_a20_gpio.c		sunxi_gpio & soc_sun7i_a20
 file	arch/arm/sunxi/sun8i_a83t_gpio.c	sunxi_gpio & soc_sun8i_a83t
 file	arch/arm/sunxi/sun8i_h3_gpio.c		sunxi_gpio & soc_sun8i_h3
+file	arch/arm/sunxi/sun8i_v3s_gpio.c		sunxi_gpio & soc_sun8i_v3s
 file	arch/arm/sunxi/sun9i_a80_gpio.c		sunxi_gpio & soc_sun9i_a80
 file	arch/arm/sunxi/sun50i_a64_gpio.c	sunxi_gpio & soc_sun50i_a64
 file	arch/arm/sunxi/sun50i_h6_gpio.c		sunxi_gpio & soc_sun50i_h6
Index: arm/sunxi/sunxi_de2_ccu.c
===================================================================
RCS file: /cvsroot/src/sys/arch/arm/sunxi/sunxi_de2_ccu.c,v
retrieving revision 1.7
diff -b -u -r1.7 sunxi_de2_ccu.c
--- arm/sunxi/sunxi_de2_ccu.c	27 Jan 2021 03:10:20 -0000	1.7
+++ arm/sunxi/sunxi_de2_ccu.c	14 Jun 2022 10:13:01 -0000
@@ -100,6 +100,8 @@
 static const struct device_compatible_entry compat_data[] = {
 	{ .compat = "allwinner,sun8i-h3-de2-clk",
 	  .data = &sun8i_h3_de2_config },
+	{ .compat = "allwinner,sun8i-v3s-de2-clk",
+	  .data = &sun8i_h3_de2_config },
 	{ .compat = "allwinner,sun50i-a64-de2-clk",
 	  .data = &sun50i_a64_de2_config },
 	{ .compat = "allwinner,sun50i-h5-de2-clk",
Index: arm/sunxi/sunxi_drm.c
===================================================================
RCS file: /cvsroot/src/sys/arch/arm/sunxi/sunxi_drm.c,v
retrieving revision 1.24
diff -b -u -r1.24 sunxi_drm.c
--- arm/sunxi/sunxi_drm.c	21 Apr 2022 21:22:25 -0000	1.24
+++ arm/sunxi/sunxi_drm.c	14 Jun 2022 10:13:01 -0000
@@ -69,6 +69,7 @@
 
 static const struct device_compatible_entry compat_data[] = {
 	{ .compat = "allwinner,sun8i-h3-display-engine" },
+	{ .compat = "allwinner,sun8i-v3s-display-engine" },
 	{ .compat = "allwinner,sun50i-a64-display-engine" },
 	DEVICE_COMPAT_EOL
 };
Index: arm/sunxi/sunxi_emac.c
===================================================================
RCS file: /cvsroot/src/sys/arch/arm/sunxi/sunxi_emac.c,v
retrieving revision 1.35
diff -b -u -r1.35 sunxi_emac.c
--- arm/sunxi/sunxi_emac.c	31 Dec 2021 14:25:22 -0000	1.35
+++ arm/sunxi/sunxi_emac.c	14 Jun 2022 10:13:01 -0000
@@ -134,6 +134,7 @@
 static const struct device_compatible_entry compat_data[] = {
 	{ .compat = "allwinner,sun8i-a83t-emac",	.value = EMAC_A83T },
 	{ .compat = "allwinner,sun8i-h3-emac",		.value = EMAC_H3 },
+	{ .compat = "allwinner,sun8i-v3s-emac",		.value = EMAC_H3 },
 	{ .compat = "allwinner,sun50i-a64-emac",	.value = EMAC_A64 },
 	{ .compat = "allwinner,sun50i-h6-emac",		.value = EMAC_H6 },
 	DEVICE_COMPAT_EOL
Index: arm/sunxi/sunxi_gpio.c
===================================================================
RCS file: /cvsroot/src/sys/arch/arm/sunxi/sunxi_gpio.c,v
retrieving revision 1.37
diff -b -u -r1.37 sunxi_gpio.c
--- arm/sunxi/sunxi_gpio.c	7 Aug 2021 16:18:45 -0000	1.37
+++ arm/sunxi/sunxi_gpio.c	14 Jun 2022 10:13:01 -0000
@@ -111,6 +111,10 @@
 	{ .compat = "allwinner,sun8i-h3-r-pinctrl",
 	  .data = &sun8i_h3_r_padconf },
 #endif
+#ifdef SOC_SUN8I_V3S
+	{ .compat = "allwinner,sun8i-v3s-pinctrl",
+	  .data = &sun8i_v3s_padconf },
+#endif
 #ifdef SOC_SUN9I_A80
 	{ .compat = "allwinner,sun9i-a80-pinctrl",
 	  .data = &sun9i_a80_padconf },
Index: arm/sunxi/sunxi_gpio.h
===================================================================
RCS file: /cvsroot/src/sys/arch/arm/sunxi/sunxi_gpio.h,v
retrieving revision 1.9
diff -b -u -r1.9 sunxi_gpio.h
--- arm/sunxi/sunxi_gpio.h	27 May 2019 23:27:14 -0000	1.9
+++ arm/sunxi/sunxi_gpio.h	14 Jun 2022 10:13:01 -0000
@@ -79,6 +79,10 @@
 extern const struct sunxi_gpio_padconf sun8i_h3_r_padconf;
 #endif
 
+#ifdef SOC_SUN8I_V3S
+extern const struct sunxi_gpio_padconf sun8i_v3s_padconf;
+#endif
+
 #ifdef SOC_SUN9I_A80
 extern const struct sunxi_gpio_padconf sun9i_a80_padconf;
 extern const struct sunxi_gpio_padconf sun9i_a80_r_padconf;
Index: arm/sunxi/sunxi_lcdc.c
===================================================================
RCS file: /cvsroot/src/sys/arch/arm/sunxi/sunxi_lcdc.c,v
retrieving revision 1.14
diff -b -u -r1.14 sunxi_lcdc.c
--- arm/sunxi/sunxi_lcdc.c	19 Dec 2021 11:01:10 -0000	1.14
+++ arm/sunxi/sunxi_lcdc.c	14 Jun 2022 10:13:01 -0000
@@ -109,6 +109,7 @@
 
 static const struct device_compatible_entry compat_data[] = {
 	{ .compat = "allwinner,sun8i-h3-tcon-tv",	.value = TYPE_TCON1 },
+	{ .compat = "allwinner,sun8i-v3s-tcon",		.value = TYPE_TCON0 },
 	{ .compat = "allwinner,sun50i-a64-tcon-lcd",	.value = TYPE_TCON0 },
 	{ .compat = "allwinner,sun50i-a64-tcon-tv",	.value = TYPE_TCON1 },
 	DEVICE_COMPAT_EOL
Index: arm/sunxi/sunxi_mixer.c
===================================================================
RCS file: /cvsroot/src/sys/arch/arm/sunxi/sunxi_mixer.c,v
retrieving revision 1.18
diff -b -u -r1.18 sunxi_mixer.c
--- arm/sunxi/sunxi_mixer.c	19 Dec 2021 11:01:10 -0000	1.18
+++ arm/sunxi/sunxi_mixer.c	14 Jun 2022 10:13:01 -0000
@@ -203,6 +203,8 @@
 static const struct device_compatible_entry compat_data[] = {
 	{ .compat = "allwinner,sun8i-h3-de2-mixer-0",
 	  .data = &mixer0_data },
+	{ .compat = "allwinner,sun8i-v3s-de2-mixer",
+	  .data = &mixer0_data },
 	{ .compat = "allwinner,sun50i-a64-de2-mixer-0",
 	  .data = &mixer0_data },
 	{ .compat = "allwinner,sun50i-a64-de2-mixer-1",
Index: arm/sunxi/sunxi_platform.c
===================================================================
RCS file: /cvsroot/src/sys/arch/arm/sunxi/sunxi_platform.c,v
retrieving revision 1.45
diff -b -u -r1.45 sunxi_platform.c
--- arm/sunxi/sunxi_platform.c	31 Jul 2021 11:34:40 -0000	1.45
+++ arm/sunxi/sunxi_platform.c	14 Jun 2022 10:13:01 -0000
@@ -490,6 +490,7 @@
 
 ARM_PLATFORM(sun8i_h2plus, "allwinner,sun8i-h2-plus", &sun8i_platform);
 ARM_PLATFORM(sun8i_h3, "allwinner,sun8i-h3", &sun8i_platform);
+ARM_PLATFORM(sun8i_v3s, "allwinner,sun8i-v3s", &sun8i_platform);
 
 static const struct arm_platform sun8i_a83t_platform = {
 	.ap_devmap = sun8i_a83t_platform_devmap,
Index: arm/sunxi/sunxi_usbphy.c
===================================================================
RCS file: /cvsroot/src/sys/arch/arm/sunxi/sunxi_usbphy.c,v
retrieving revision 1.16
diff -b -u -r1.16 sunxi_usbphy.c
--- arm/sunxi/sunxi_usbphy.c	27 Jan 2021 03:10:20 -0000	1.16
+++ arm/sunxi/sunxi_usbphy.c	14 Jun 2022 10:13:01 -0000
@@ -91,6 +91,7 @@
 	{ .compat = "allwinner,sun7i-a20-usb-phy",	.value = USBPHY_A20 },
 	{ .compat = "allwinner,sun8i-a83t-usb-phy",	.value = USBPHY_A83T },
 	{ .compat = "allwinner,sun8i-h3-usb-phy",	.value = USBPHY_H3 },
+	{ .compat = "allwinner,sun8i-v3s-usb-phy",	.value = USBPHY_H3 },
 	{ .compat = "allwinner,sun50i-a64-usb-phy",	.value = USBPHY_A64 },
 	{ .compat = "allwinner,sun50i-h6-usb-phy",	.value = USBPHY_H6 },
 	DEVICE_COMPAT_EOL
Index: evbarm/conf/GENERIC
===================================================================
RCS file: /cvsroot/src/sys/arch/evbarm/conf/GENERIC,v
retrieving revision 1.107
diff -b -u -r1.107 GENERIC
--- evbarm/conf/GENERIC	2 Apr 2022 11:16:07 -0000	1.107
+++ evbarm/conf/GENERIC	14 Jun 2022 10:13:01 -0000
@@ -26,6 +26,7 @@
 options 	SOC_SUN7I_A20
 options 	SOC_SUN8I_A83T
 options 	SOC_SUN8I_H3
+options 	SOC_SUN8I_V3S
 options 	SOC_SUN9I_A80
 options 	SOC_TEGRA124
 options 	SOC_VIRT
